File: lambda_function.js
const jwt = require('jsonwebtoken');
const jwkToPem = require('jwk-to-pem');
const https = require('https');

const COGNITO_POOL_URL = process.env.COGNITO_POOL_URL; 
const COGNITO_POOL_PUBLIC_KEY_URL = `${COGNITO_POOL_URL}/.well-known/jwks.json`;

exports.lambda_handler = async (event) => {
    const authHeader = event.headers.Authorization || event.headers.authorization; // Headers may be case-sensitive

    if (!authHeader) {
        return {
            statusCode: 401,
            body: JSON.stringify({ message: 'Authorization header is missing' })
        };
    }   

    // Extract the token from the Authorization header
    const token = authHeader.split(' ')[1]; // Assumes "Bearer YOUR_TOKEN_HERE" format

    // Now, you can use the token for your logic, like verifying it
    //console.log('Extracted Token:', token);
    
    // Check if the token is blank, null, or undefined
    // if (!token || token.trim() === '' || typeof token === 'undefined') {
    //     return {
    //         statusCode: 401,
    //         body: JSON.stringify({ 
    //             error: 'Token is blank, null, or undefined',
    //             token: token
    //         })
    //     };
    // }
    
    try {
        const jwks = await getJWKS();
        const pem = jwkToPem(jwks.keys[0]); // Convert the JWK to PEM format
        const decodedToken = jwt.verify(token, pem, { algorithms: ['RS256'] });

        return {
            statusCode: 200,
            body: JSON.stringify(decodedToken)
        };
    } catch (err) {
        console.error('Failed to decode or verify token:', err);
        return {
            statusCode: 401,
            body: JSON.stringify({ 
                error: 'Invalid token or unauthorized',
                token: token  // Adding the token to the error response
            })
        };
    }
};

function getJWKS() {
    return new Promise((resolve, reject) => {
        https.get(COGNITO_POOL_PUBLIC_KEY_URL, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                resolve(JSON.parse(data));
            });

        }).on('error', (err) => {
            reject(err);
        });
    });
}

File: lambda_function.py
import jwt
import requests
import json
import os
from jwt.algorithms import RSAAlgorithm

# Get Cognito pool URL and audience from Lambda environment variables
COGNITO_POOL_URL = os.environ['COGNITO_POOL_URL']
COGNITO_APP_CLIENT_ID = os.environ['COGNITO_APP_CLIENT_ID']
COGNITO_POOL_PUBLIC_KEY_URL = COGNITO_POOL_URL + '/.well-known/jwks.json'


def lambda_handler(event, context):
    token = event['userToken']  # Assuming the token is passed in the event
    
    # Get Cognito's public keys
    keys = requests.get(COGNITO_POOL_PUBLIC_KEY_URL).json()['keys']
    
    # Try to decode the token using each key
    decoded_token = None
    for key in keys:
        try:
            decoded_token = jwt.decode(
                token,
                RSAAlgorithm.from_jwk(json.dumps(key)),
                algorithms=['RS256'],
                audience=COGNITO_APP_CLIENT_ID,
                issuer=COGNITO_POOL_URL
            )
            break
        except jwt.DecodeError:
            continue

    if decoded_token:
        user_id = decoded_token['sub']  # 'sub' claim is the standard claim for the unique user identifier in Cognito
        return {'userId': user_id}
    else:
        raise Exception('Invalid token or unauthorized')


